File: mathewDavidov/assignments/treesBFS/lc102/lc102.java
/**
 * Problem Statement:
 * Given a binary tree, populate an array to represent its level-by-level traversal. 
 * You should populate the values of all nodes of each level from left to right in separate sub-arrays.
 * 
 * Example:
 *     4
 *    3 7
 *  1 2 6 10
 * = [[4], [3, 7], [1, 2, 6, 10]]
 * 
 * Edge cases:
 * null root -> return
 * 
 * initialize a queue and container list
 * add the root to the queue
 * while the queue is not empty:
 *   initialize the size of the current width of level
 *   initialize a list for the current level
 *   loop from 0 to the current level size:
 *     poll a node from the queue
 *     if the left child exists, add it the queue
 *     if the right child exists, add it to the queue
 *     add the value of the current node to the list
 *   add the current level list to the container list
 * return the container list
 */

import java.util.*;

public class lc102 {
    public static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> levelOrder = new ArrayList<>();

        if (root == null) {
            return levelOrder;
        }

        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> levelList = new ArrayList<>();

            for (int i = 0; i < levelSize; i++) {
                TreeNode current = queue.poll();

                if (current.left != null) {
                    queue.offer(current.left);
                }

                if (current.right != null) {
                    queue.offer(current.right);
                }

                levelList.add(current.val);
            }

            levelOrder.add(levelList);
        }

        return levelOrder;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(12);
        root.left = new TreeNode(7);
        root.right = new TreeNode(1);
        root.left.left = new TreeNode(9);
        root.right.left = new TreeNode(10);
        root.right.right = new TreeNode(5);
        List<List<Integer>> result = levelOrder(root);
        System.out.println("Level order traversal: ");
        System.out.println(result);
    }
}

class TreeNode {
    public int val = 0;
    public TreeNode left;
    public TreeNode right;
   
    public TreeNode(int val) {
       this.val = val;
       this.left = this.right = null;
    }
}

File: alanEappen/assignments/treesBFS/lc102/lc102.js
// Problem Statement #

// Given a binary tree, populate an array to represent its level-by-level traversal. You should populate the values of all nodes of each level from left to right in separate sub-arrays.

class TreeNode {

    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
};

//Given: BinaryTree with root node
//Goal: return an array that has seperate sub arrays from left to right 

//first declare an array as queue that contains the root node 
//variable 
//while(queue.length) initiate while loop if the length exists

//Declaring a level to push
//let currLevel =[]
//Loop through the level to inlcude all the nodes at the current level
//for(let i = 0; i<queue.length; i++)
// currentNode is equal to next node

//if condition (currentNode.left)
//queue.push(currentNode.left)

//if condition (currentNode.right)
//queue.push(currentNode.right)

//after adding left and right we push the current lvl

// once that level is finished
//push currentlevel to level
//return level
//Edge cases:
//if the root node is null
//return []


const traverse = function (root) {
    let levels = [];
    let queue = []


    if (!root) {
        return []
    }
    queue.push(root)
    while (queue.length) {
        let currLevel = []
        let len = queue.length
        for (let i = 0; i < len; i++) {
            let currentNode = queue.shift();

            if (currentNode.left)
                queue.push(currentNode.left)

            if (currentNode.right)
                queue.push(currentNode.right)

            currLevel.push(currentNode.value)
        }
        levels.push(currLevel)
    }
    return levels;

};



var root = new TreeNode(12);                                      //[12]
root.left = new TreeNode(7);                                    // [7] [1]
root.right = new TreeNode(1);                                  //[9] [10] [5]
root.left.left = new TreeNode(9);
root.right.left = new TreeNode(10);
root.right.right = new TreeNode(5);
console.log(`Level order traversal: ${traverse(root)}`); //Output [[12],[7,1],[9,10,5]]


  // Solution
  // -----

  // const traverse = function(root) {
  //   // If root is null return an empty array
  //   if(!root) return []

  //   const queue = [root] // initialize the queue with root
  //   const levels = [] // declare output array

  //   while(queue.length !== 0){
  //      const queueLength = queue.length // Get the length prior to dequeueing
  //      const currLevel = [] // Declare this level
  //      // loop through to exhaust all options and only to include nodes at currLevel
  //      for(let i = 0; i < queueLength; i++){
  //          // Get next node
  //          const current = queue.shift()

  //          if(current.left){
  //              queue.push(current.left)
  //          }
  //          if(current.right){
  //              queue.push(current.right)
  //          }
  //          // After we add left and right for current, we add to currLevel
  //          currLevel.push(current.val)
  //      }
  //      // Level has been finished. Push into output array
  //      levels.push(currLevel)
  //  }
  //   return levels
  // }

  // -----

  // Time complexity #
  // The time complexity of the above algorithm is O(N), where ‘N’ is the total number of nodes in the tree. This is due to the fact that we traverse each node once.

  // Space complexity #
  // The space complexity of the above algorithm will be O(N) as we need to return a list containing the level order traversal. We will also need O(N) space for the queue. Since we can have a maximum of N/2 nodes at any level (this could happen only at the lowest level), therefore we will need O(N) space to store them in the queue.
